### PR TITLE
reduce extension method for cross-assembly case

### PR DESCRIPTION
**Pull request by ansyral**
@chenkennt @vwxyzh @hellosnow @superyyrrzz @qinezh

### CODE REVIEW

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                name = name.Remove(index);
+            }
+            var members = GetMembersCore(container, name);
+            if (members.Length == 0)
+            {
+                return null;
+            }
+            else if (members.Length > 1)
+            {
+                Debug.Assert(false, "Found multiple members of specified name:\r\n" + string.Join("\r\n", members));
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs]: use Debug.Fail?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs]: add qualifiedName into error msg?

**Comment by vwxyzh** [comment]
👍

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        public static ISymbol GetMember(this INamespaceOrTypeSymbol container, string qualifiedName)
+        {
+            var name = TemplateParameterRegex.Replace(qualifiedName, string.Empty);
+            var index = name.IndexOf('(');
+            if (index != -1)
+            {
+                name = name.Remove(index);
+            }
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs]: move into SplitQualifiedName?

**Comment by superyyrrzz** [comment]
> namespace Microsoft.DocAsCode.Metadata.ManagedReference

add copyright header

---
Refers to: src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs:1 in 833099e. [](commit_id = 833099eee2578d63873c747b651c3b830c40fa4f, deletion_comment = False)

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        private static IEnumerable<string> SplitQualifiedName(
+            string pstrName)
+        {
+            Debug.Assert(pstrName != null);
+            do
+            {
+                // If we see consecutive dots, the second is part of the method name
+                // (i.e. ".ctor" or ".cctor").
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs]: `.cctor` => `..ctor` ?

**Comment by ansyral** [comment]
> namespace Microsoft.DocAsCode.Metadata.ManagedReference

sorry i didn't notice your comment, will fix it in next iteration

---
In reply to: [226365356](https://github.com/dotnet/docfx/pull/385#issuecomment-226365356) [](ancestors = 226365356)

---
Refers to: src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs:1 in 833099e. [](commit_id = 833099eee2578d63873c747b651c3b830c40fa4f, deletion_comment = False)

**Comment by superyyrrzz** [comment]
:+1:

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                    }
+                }
+            }
+        }
+
+        private static List<string> GetQualifiedNameList(ISymbol symbol)
+        {
+            var names = new List<string>();
+            var current = symbol;
+            while ((current as INamespaceSymbol)?.IsGlobalNamespace != true)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs]: remove and prepend with ‘!’？

Comment (ansyral) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/CodeAnalysisSymbolExtensions.cs]: the type of ?. is nullable<T>

**Comment by hellosnow** [comment]
:shipit:

**Pull request merged**